File: apps/mobile/lib/core/logic/price_history_series.dart
import '../models/app_settings.dart';
import '../models/card_model.dart';
import 'pricing.dart';

/// Visible window for the Low chart. Everyone defaults to [last30]; Pro may
/// opt in to [full] for one visit.
enum PriceHistoryWindow { last30, full }

/// One observed Low for the selected marketplace on a calendar day.
class LowObservation {
  const LowObservation({required this.date, required this.low});

  /// Date-only (`year/month/day`); time-of-day is ignored.
  final DateTime date;
  final double low;
}

/// Derived Low series for the history chart. [PricePoint] stays the snapshot;
/// this type names observed Low vs the clipped window vs the change summary.
class PriceHistorySeries {
  const PriceHistorySeries({
    required this.window,
    required this.windowStart,
    required this.points,
    required this.hasOlder,
    required this.isPro,
  });

  final PriceHistoryWindow window;
  final DateTime windowStart;
  final List<LowObservation> points;
  final bool hasOlder;
  final bool isPro;

  bool get chartable => points.length >= 2;

  /// Last visible Low minus first visible Low. Null when not [chartable].
  double? get delta =>
      chartable ? points.last.low - points.first.low : null;

  bool get showProCta => !isPro && hasOlder && chartable;

  bool get showSpanControl => isPro && hasOlder && chartable;

  String? changeLabel(Pricing pricing) {
    final d = delta;
    if (d == null) return null;
    if (d == 0) return 'Low unchanged';
    final amount = pricing.formatValue(d.abs());
    return d > 0 ? 'Low up $amount' : 'Low down $amount';
  }

  static DateTime dateOnly(DateTime d) => DateTime(d.year, d.month, d.day);

  /// Inclusive start of the 30-calendar-day window (today minus 29 days).
  static DateTime windowStartFor(DateTime now) =>
      dateOnly(now).subtract(const Duration(days: 29));

  static double? _lowFor(PricePoint snap, PriceSource source) {
    switch (source) {
      case PriceSource.tcgplayer:
        return snap.tcgLow;
      case PriceSource.cardmarket:
        return snap.cmLow;
    }
  }

  static List<LowObservation> observations(
    List<PricePoint> snapshots,
    PriceSource source,
  ) {
    final out = <LowObservation>[];
    for (final snap in snapshots) {
      final low = _lowFor(snap, source);
      if (low == null) continue;
      out.add(LowObservation(date: dateOnly(snap.capturedOn), low: low));
    }
    return out;
  }

  factory PriceHistorySeries.fromSnapshots({
    required List<PricePoint> snapshots,
    required PriceSource source,
    required bool isPro,
    PriceHistoryWindow window = PriceHistoryWindow.last30,
    DateTime? now,
  }) {
    final today = now ?? DateTime.now();
    final windowStart = windowStartFor(today);
    final all = observations(snapshots, source);
    final hasOlder = all.any((o) => o.date.isBefore(windowStart));
    final points = window == PriceHistoryWindow.full
        ? all
        : all.where((o) => !o.date.isBefore(windowStart)).toList();
    return PriceHistorySeries(
      window: window,
      windowStart: windowStart,
      points: points,
      hasOlder: hasOlder,
      isPro: isPro,
    );
  }
}

File: apps/web/src/utils/priceHistorySeries.js
/**
 * Observed Low series for one Printing's price-history chart.
 * Golden cases: packages/contracts/price_history_series.json
 *
 * Snapshot rows stay snapshots. This helper names the marketplace Low, the
 * clipped window, and the derived delta. Null Low is a gap, never 0.
 */

export const PRICE_HISTORY_WINDOW = {
    last30: 'last30',
    full: 'full',
};

/** Calendar day only — never UTC midnight of a YYYY-MM-DD string. */
export function dateOnly(value) {
    if (value instanceof Date) {
        return new Date(value.getFullYear(), value.getMonth(), value.getDate());
    }
    const s = String(value ?? '').slice(0, 10);
    const [year, month, day] = s.split('-').map(Number);
    return new Date(year, month - 1, day);
}

export function formatDateOnly(value) {
    const d = dateOnly(value);
    const y = d.getFullYear();
    const m = String(d.getMonth() + 1).padStart(2, '0');
    const day = String(d.getDate()).padStart(2, '0');
    return `${y}-${m}-${day}`;
}

/** Inclusive start of the 30-calendar-day window (today minus 29 days). */
export function windowStartFor(now = new Date()) {
    const today = dateOnly(now);
    return new Date(today.getFullYear(), today.getMonth(), today.getDate() - 29);
}

const lowFor = (snap, source) => {
    const raw = source === 'cardmarket' ? snap?.cm_low : snap?.tcg_low;
    if (raw === null || raw === undefined || raw === '') return null;
    const n = Number(raw);
    return Number.isFinite(n) ? n : null;
};

export function observations(snapshots, source) {
    const out = [];
    for (const snap of snapshots || []) {
        const low = lowFor(snap, source);
        if (low === null) continue;
        out.push({ date: formatDateOnly(snap.captured_on), low });
    }
    out.sort((a, b) => (a.date < b.date ? -1 : a.date > b.date ? 1 : 0));
    return out;
}

/**
 * Visible Low series for the history chart.
 *
 * @param {object} opts
 * @param {object[]} opts.snapshots rows with captured_on + nullable tcg_low / cm_low
 * @param {'tcgplayer'|'cardmarket'} opts.source
 * @param {boolean} [opts.isPro]
 * @param {'last30'|'full'} [opts.window]
 * @param {Date|string} [opts.now]
 */
export function priceHistorySeries({
    snapshots,
    source,
    isPro = false,
    window = PRICE_HISTORY_WINDOW.last30,
    now = new Date(),
} = {}) {
    const windowStart = windowStartFor(now);
    const windowStartKey = formatDateOnly(windowStart);
    const all = observations(snapshots, source);
    const hasOlder = all.some((o) => o.date < windowStartKey);
    const points = window === PRICE_HISTORY_WINDOW.full
        ? all
        : all.filter((o) => o.date >= windowStartKey);
    const chartable = points.length >= 2;
    const delta = chartable ? points[points.length - 1].low - points[0].low : null;
    return {
        window,
        windowStart: windowStartKey,
        points,
        hasOlder,
        isPro: Boolean(isPro),
        chartable,
        delta,
        showSpanControl: Boolean(isPro) && hasOlder && chartable,
    };
}

export function formatObservedLow(value, source) {
    const n = Number(value);
    if (!Number.isFinite(n)) return '—';
    if (source === 'cardmarket') return `€${n.toFixed(2)}`;
    return `$${n.toFixed(2)}`;
}

export function changeLabel(series, source) {
    if (!series?.chartable || series.delta == null) return null;
    if (series.delta === 0) return 'Low unchanged';
    const amount = formatObservedLow(Math.abs(series.delta), source);
    return series.delta > 0 ? `Low up ${amount}` : `Low down ${amount}`;
}

export function emptyHistoryCopy(snapshots, source) {
    const other = source === 'cardmarket' ? 'tcgplayer' : 'cardmarket';
    const otherChartable = priceHistorySeries({
        snapshots,
        source: other,
        isPro: true,
        window: PRICE_HISTORY_WINDOW.full,
    }).chartable;
    if (otherChartable) {
        const label = source === 'cardmarket' ? 'CardMarket' : 'TCGplayer';
        return `History isn't available for ${label} yet`;
    }
    return 'History not available yet';
}

export function formatHistoryDate(value) {
    return dateOnly(value).toLocaleDateString('en-US', {
        month: 'short',
        day: 'numeric',
    });
}
